qemu/tests/chardev_serial_login.py

```diff
@@ -0,0 +1,46 @@
+from virttest import error_context, env_process
+from avocado.utils import process
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Verify the login function of pci-serial (RHEL and x86 only):
+    1) Start guest with pci-serial with backend unix_socket
+    2) append console=ttyS0 to guest kernel config
+    3) append /dev/ttyS0 to /etc/security
+    4) reboot guest
+    5) nc host/path for unix and tcp backend.
+    6) login to guest and move files, restart network
+    7) repeat step 1 to 6 with pty, tcp backend
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    serial_id = params.objects('serials')[-1]
+    params['start_vm'] = 'yes'
+    for backend in ['unix_socket', 'tcp_socket', 'pty']:
+        params['chardev_backend_%s' % serial_id] = backend
+        vm = params['main_vm']
+        env_process.preprocess_vm(test, params, env, vm)
+        vm = env.get_vm(params["main_vm"])
+        vm.verify_alive()
+        session = vm.wait_for_login()
+        assert session.cmd('lspci | grep "PCI 16550A"') is not None
+        session.cmd('echo "console=tty0 console=ttyS0,115200" >> /boot/grub2/grub.conf')
+        session.cmd('echo "/dev/ttyS0" >> /dev/ttyS0')
+        vm.reboot()
+        vm.verify_alive()
+        serial_device = vm.devices.get(serial_id)
+        chardev_qid = serial_device.get_param("chardev")
+        chardev_device = vm.devices.get_by_qid(chardev_qid)[0]
+        if backend == 'tcp_socket':
+            process.run('nc %s %s' % (chardev_device.params['host'], chardev_device.params['port']))
+        elif backend == 'unix_socket':
+            process.run('nc -U %s' % chardev_device.params['path'])
```

Comment:
Could you try `aexpect.ShellSession`
https://github.com/avocado-framework/avocado-vt/blob/3fd00812165e47b7efc9ba8ad2bc1efa80ececd9/virttest/qemu_vm.py#L2472